### PR TITLE
jagged lr restart scheudler

### PR DESCRIPTION
**Pull request by winglian**
generic LR restart scheduler for ReLoRA and other techniques.

## Summary by CodeRabbit

* **New Features**
  * Introduced a new learning rate scheduler option supporting "jagged restarts," enabling periodic resets with customizable warmup and annealing steps during training.
  * Added new training arguments to configure jagged restart frequency, warmup steps, and anneal steps.
  * Integrated jagged restart parameters into training configuration and trainer selection.
  * Updated ReLoRA support to use jagged restart settings for step control.

* **Improvements**
  * Enhanced scheduler configuration to support advanced learning rate scheduling patterns for improved training flexibility.
  * Refined validation logic to require jagged restart parameters when using ReLoRA.

* **Removals**
  * Removed the dedicated ReLoRATrainer class and related custom scheduler, unifying functionality under the jagged restart scheduler settings.
  * Simplified ReLoRA configuration by replacing multiple step-related fields with a single enable flag.

### CODE REVIEW

File: src/axolotl/core/training_args.py [hunk 1]
```diff
@@ -76,6 +76,22 @@ class AxolotlTrainingMixins:
...
+        metadata={
+            "help": "how many warmup steps to take after reset for jagged restarts"
+        },
+    )
+    jagged_restarts_anneal_steps: Optional[int] = field(
+        default=None,
+        metadata={
+            "help": "how many anneal steps to take before reset for jagged restarts"
+        },
+    )
```
Comment (NanoCode012) [inline src/axolotl/core/training_args.py]: Could we add to `config.qmd`?

File: src/axolotl/core/trainers/base.py [hunk 2]
```diff
@@ -143,7 +144,21 @@ def create_scheduler(
...
+                    anneal_steps = (
+                        self.args.jagged_restarts_anneal_steps or 1
+                    )
+                    self.lr_scheduler = JaggedLRRestartScheduler(  # pylint: disable=attribute-defined-outside-init
+                        optimizer,
+                        self.lr_scheduler,
+                        self.args.jagged_restart_steps,
+                        warmup_steps,
+                        anneal_steps,
+                    )
```
Comment (NanoCode012) [inline src/axolotl/core/trainers/base.py]: - Do we need to get the return from `super().create_scheduler`?

- Would it make more sense to move this section one tab left as it could apply for other lr schedulers too?

File: src/axolotl/core/training_args.py (hunk 1 above)
Comment (winglian) [reply src/axolotl/core/training_args.py]: shouldn't be necessary now with auto-doc

File: src/axolotl/core/builders/causal.py [hunk 3]
```diff
@@ -271,20 +268,34 @@ def build(self, total_num_steps):
...
+                self.cfg.jagged_restart_steps
+            )
+            if self.cfg.jagged_restart_warmup_steps:
+                training_arguments_kwargs["jagged_restart_warmup_steps"] = (
+                    self.cfg.jagged_restart_warmup_steps
+                )
+            if self.cfg.jagged_restart_anneal_steps:
+                training_arguments_kwargs["jagged_restart_anneal_steps"] = (
+                    self.cfg.jagged_restart_anneal_steps
+                )
```
Comment (NanoCode012) [inline src/axolotl/core/builders/causal.py]: Confused on the difference between the two. Is jagged the new method? Why do we have both here but remove relora from input schema?

File: src/axolotl/utils/schemas/peft.py [hunk 4]
```diff
@@ -187,18 +187,10 @@ def validate_lora_dropout(cls, data):
...
-        json_schema_extra={"description": "Number of per-restart warmup steps"},
-    )
-    relora_anneal_steps: int | None = Field(
+    relora: bool | None = Field(
         default=None,
         json_schema_extra={
-            "description": "Number of anneal steps for each relora cycle"
+            "description": "Whether to use ReLoRA. Use with jagged_restart_*steps options."
         },
     )
```
Comment (NanoCode012) [inline src/axolotl/utils/schemas/peft.py:195]: We need to add deprecation error and migration steps before we remove this from schema?

File: src/axolotl/core/builders/causal.py (hunk 3 above)
Comment (winglian) [reply src/axolotl/core/builders/causal.py]: We only have the user define the jagged, the relora args are for the callback handler

File: src/axolotl/utils/schemas/peft.py (hunk 4 above)
Comment (winglian) [reply src/axolotl/utils/schemas/peft.py:195]: I think this is a minor enough feature that we can skip that for this one

**Review by NanoCode012 — APPROVED** [review]
I think this looks good
